x/gov/tally_test.go

```diff
@@ -0,0 +1,366 @@
+	stakeHandler(ctx, val3CreateMsg)
+
+	proposal := keeper.NewProposal(ctx, "Test", "description", "Text")
+	proposalID := proposal.ProposalID
+	proposal.Status = "VotingPeriod"
+	keeper.SetProposal(ctx, proposal)
+
+	err := keeper.AddVote(ctx, proposalID, addrs[0], "Abstain")
+	assert.Nil(t, err)
+	err = keeper.AddVote(ctx, proposalID, addrs[1], "Yes")
+	assert.Nil(t, err)
+	err = keeper.AddVote(ctx, proposalID, addrs[2], "No")
+	assert.Nil(t, err)
+
+	passes, _ := tally(ctx, keeper, keeper.GetProposal(ctx, proposalID))
+
+	assert.False(t, passes)
+}
+
+func TestTallyOnlyValidatorsNonVoter(t *testing.T) {
+	ctx, _, keeper := createTestInput(t, false, 100)
+	stakeHandler := stake.NewHandler(keeper.sk)
+
+	dummyDescription := stake.NewDescription("T", "E", "S", "T")
+	val1CreateMsg := stake.NewMsgCreateValidator(addrs[0], crypto.GenPrivKeyEd25519().PubKey(), sdk.Coin{"steak", 6}, dummyDescription)
+	stakeHandler(ctx, val1CreateMsg)
+	val2CreateMsg := stake.NewMsgCreateValidator(addrs[1], crypto.GenPrivKeyEd25519().PubKey(), sdk.Coin{"steak", 6}, dummyDescription)
+	stakeHandler(ctx, val2CreateMsg)
+	val3CreateMsg := stake.NewMsgCreateValidator(addrs[2], crypto.GenPrivKeyEd25519().PubKey(), sdk.Coin{"steak", 7}, dummyDescription)
+	stakeHandler(ctx, val3CreateMsg)
+
+	proposal := keeper.NewProposal(ctx, "Test", "description", "Text")
+	proposalID := proposal.ProposalID
+	proposal.Status = "VotingPeriod"
+	keeper.SetProposal(ctx, proposal)
+
+	err := keeper.AddVote(ctx, proposalID, addrs[1], "Yes")
+	assert.Nil(t, err)
+	err = keeper.AddVote(ctx, proposalID, addrs[2], "No")
+	assert.Nil(t, err)
+
+	passes, nonVoting := tally(ctx, keeper, keeper.GetProposal(ctx, proposalID))
+
+	assert.False(t, passes)
+	assert.Equal(t, 1, len(nonVoting))
+	assert.Equal(t, addrs[0], nonVoting[0])
+}
+
+func TestTallyDelgatorOverride(t *testing.T) {
+
+	ctx, _, keeper := createTestInput(t, false, 100)
+	stakeHandler := stake.NewHandler(keeper.sk)
+
+	dummyDescription := stake.NewDescription("T", "E", "S", "T")
+	val1CreateMsg := stake.NewMsgCreateValidator(addrs[0], crypto.GenPrivKeyEd25519().PubKey(), sdk.Coin{"steak", 5}, dummyDescription)
+	stakeHandler(ctx, val1CreateMsg)
+	val2CreateMsg := stake.NewMsgCreateValidator(addrs[1], crypto.GenPrivKeyEd25519().PubKey(), sdk.Coin{"steak", 6}, dummyDescription)
+	stakeHandler(ctx, val2CreateMsg)
+	val3CreateMsg := stake.NewMsgCreateValidator(addrs[2], crypto.GenPrivKeyEd25519().PubKey(), sdk.Coin{"steak", 7}, dummyDescription)
+	stakeHandler(ctx, val3CreateMsg)
+
+	delegator1Msg := stake.NewMsgDelegate(addrs[3], addrs[2], sdk.Coin{"steak", 30})
+	stakeHandler(ctx, delegator1Msg)
+
+	proposal := keeper.NewProposal(ctx, "Test", "description", "Text")
+	proposalID := proposal.ProposalID
+	proposal.Status = "VotingPeriod"
+	keeper.SetProposal(ctx, proposal)
+
+	err := keeper.AddVote(ctx, proposalID, addrs[0], "Yes")
```

Comment:
consider abstracting this call and the assert nil into another wrapper method in the test file. (Only because alot of lines are being spent on assert.Nil(t,err) in this file for keeper.AddVote calls, which breaks readability a bit)

Reply:
Just to hide the assert Nils?  Idk, I personally like to see exactly what's being asserted in a single flow of a test, instead of jumping between functions.